// angular8-yandex-maps/esm2020/lib/components/ya-control/ya-control.directive.mjs
import { Directive, EventEmitter, Input, Output, } from '@angular/core';
import { Subject, takeUntil } from 'rxjs';
import { filter, take } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "../ya-map/ya-map.component";
/**
 * The `ya-control` component wraps `ymaps.control[YaControlType]` classes from the Yandex.Maps API.
 * You can configure `ymaps.control[YaControlType]` via the component's inputs.
 * API Events can be bound only manually. You can use `ready` event to get an instance.
 *
 * <example-url>https://stackblitz.com/edit/searchcontrol?embed=1&view=preview</example-url>
 *
 * @example
 * ```html
 * <ya-map [center]="[55.761952, 37.620739]">
 *   <ya-control
 *     type="RoutePanel"
 *     [parameters]="{ options: { float: 'right' } }"
 *   ></ya-control>
 * </ya-map>
 * ```
 */
export class YaControlDirective {
    constructor(ngZone, yaMapComponent) {
        this.ngZone = ngZone;
        this.yaMapComponent = yaMapComponent;
        this.destroy$ = new Subject();
        /**
         * Control instance is added to a Map.
         */
        this.ready = new EventEmitter();
    }
    ngOnChanges() {
        if (this.control) {
            console.warn('Control does not support dynamic configuration. You can config it manually using ymaps or recreate the component with new configuration');
        }
    }
    ngOnInit() {
        this.yaMapComponent.map$
            .pipe(filter(Boolean), take(1), takeUntil(this.destroy$))
            .subscribe((map) => {
            const control = new ymaps.control[this.type](this.parameters);
            this.control = control;
            // RoutePanel ignores state in parameters. API bug
            if (control instanceof ymaps.control.RoutePanel &&
                this.parameters &&
                this.parameters.state) {
                control.routePanel.state.set({ ...this.parameters.state });
            }
            map.controls.add(control);
            this.ngZone.run(() => this.ready.emit({ ymaps, target: control }));
        });
    }
    ngOnDestroy() {
        if (this.control) {
            this.yaMapComponent?.map$.value?.controls.remove(this.control);
        }
        this.destroy$.next();
        this.destroy$.complete();
    }
}
YaControlDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.5", ngImport: i0, type: YaControlDirective, deps: [{ token: i0.NgZone }, { token: i1.YaMapComponent }], target: i0.ɵɵFactoryTarget.Directive });
YaControlDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.2.5", type: YaControlDirective, selector: "ya-control", inputs: { type: "type", parameters: "parameters" }, outputs: { ready: "ready" }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.5", ngImport: i0, type: YaControlDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ya-control',
                }]
        }], ctorParameters: function () { return [{ type: i0.NgZone }, { type: i1.YaMapComponent }]; }, propDecorators: { type: [{
                type: Input
            }], parameters: [{
                type: Input
            }], ready: [{
                type: Output
            }] } });
